File: Stack/java codes/valid parentheses.java
class Solution {
  // This function checks if the given string is balanced.
  public boolean isValid(String s) {
    // Create a stack to store the opening brackets.
    Stack<Character> stack = new Stack<Character>();

    // Iterate through the string.
    for (char c : s.toCharArray()) {
      // If the current character is an opening bracket, push it onto the stack.
      if (c == '(' || c == '{' || c == '[') {
        stack.push(c);
      }

      // If the current character is a closing bracket, check if it matches the top of the stack.
      else if (stack.isEmpty() || (c == '}' && stack.peek() != '{') || (c == ')' && stack.peek() != '(') || (c == ']' && stack.peek() != '[')) {
        // If the brackets do not match, return false.
        return false;
      }

      // Otherwise, pop the top of the stack.
      else {
        stack.pop();
      }
    }

    // If the stack is empty, return true. This means that all of the brackets were balanced.
    return stack.isEmpty();
  }
}

File: Stack/Easy/valid parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        std::stack<char> st;  // create a stack to store opening brackets
        for(auto itr = s.begin(); itr!= s.end(); ++itr){ // iterate through each character in the string

            if(*itr == '(' || *itr =='{'|| *itr == '[')  // if it is an opening bracket, push it onto the stack
                st.push(*itr);

            else if(st.empty() || st.top() == '{' && *itr != '}'   || st.top() == '(' && *itr != ')' || st.top() == '[' && *itr != ']'){ // if it is a closing bracket, check if it matches the corresponding opening bracket at the top of the stack

                return false;  // if it does not match, return false
            }

            else st.pop();  // if it matches, pop the opening bracket from the stack

        }

        return st.empty() ? true : false;  // check if the stack is empty. if it is, the string is valid. otherwise, it is invalid.
    }
};
